### PR TITLE
lxd/storage: Enable LO_FLAGS_DIRECT_IO by default for loop devices

### LINKED ISSUE #9407: Enable LO_FLAGS_DIRECT_IO by default on loopback devices
Turning on `LO_FLAGS_DIRECT_IO` is recommended by hch and axeboe and will improve performance of loop devices singificantly. We should turn this on.

### PR DESCRIPTION
**Pull request by archevan**
Resolves #9407

My current implementation passes the LO_FLAGS_DIRECT_IO flag into the PrepareLoopDev function when it's called from the lvm and btrfs drivers. I also included C functions to manually set direct IO on a given loop device if needed instead.

Also, should I add a storage pool configuration option to allow users to disable direct IO?

### CODE REVIEW

File: lxd/storage/drivers/driver_lvm_utils.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ func (d *lvm) openLoopFile(source string) (*os.File, error) {
 		defer unlock()
 
 		// Try to prepare new loop device.
-		loopF, err := PrepareLoopDev(source, 0)
+		loopF, err := PrepareLoopDev(source, LoFlagsDirectIO)
```
Comment (stgraber) [inline lxd/storage/drivers/driver_lvm_utils.go:63]: Should be split into its own commit.

File: lxd/storage/drivers/utils_cgo.go [hunk 2]
```diff
@@ -348,3 +360,25 @@ func UnsetAutoclearOnLoopDev(loopFd int) error {
...
+func UnsetDirectIOLoopDev(loopFd int) error {
+	ret, err := C.unset_direct_io_loop_device(C.int(loopFd))
+	if ret < 0 {
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("Failed to unset LO_FLAGS_DIRECT_IO")
+	}
+	return nil
+}
```
Comment (stgraber) [inline lxd/storage/drivers/utils_cgo.go:384]: Those two aren't used anywhere, so they should be safe to drop (and the cgo code along with them).

File: lxd/storage/drivers/utils_cgo.go [hunk 3]
```diff
@@ -265,6 +276,7 @@ import "C"
 // LoFlagsAutoclear determines whether the loop device will autodestruct on last
 // close.
 const LoFlagsAutoclear int = C.LO_FLAGS_AUTOCLEAR
+const LoFlagsDirectIO int = C.LO_FLAGS_DIRECT_IO
```
Comment (stgraber) [inline lxd/storage/drivers/utils_cgo.go:279]: So one catch with `LO_FLAGS_DIRECT_IO` is that it's only supported on Linux 4.10 and above.

As a result, I think we'd need `prepare_loop_dev` to check if `LO_FLAGS_DIRECT_IO` is part of its `flags` and if it is then handle the `ioctl` failure which may come if the kernel doesn't support that flag, then strip the flag and retry without it.

@brauner do you know what the `ioctl` would return if unsupported? ENOSYS? EINVAL?

**Comment by stgraber** [comment]
So looking into the systemd code linked in the issue, you can find:
https://github.com/systemd/systemd/pull/21044/files#diff-e20c8b3ba10a6099dd4cf12f9b95ade0cba34523b5662ac7779b11f92c37da41R482

Which has the logic to properly handle that flag. Looks like it's slightly more involved than done in this PR as we need to:
 - Make sure that the file backing the loop is open with O_DIRECT as well, if we fail to do that, then we should strip the ioctl flag
 - Do a separate ioctl call to set DIRECT_IO mode on the loop
 - Handle failure of that ioctl as the kernel may not be supported by the kernel

@brauner does that sound right?

File: lxd/storage/drivers/utils_cgo.go (hunk 3 above)
Comment (archevan) [reply lxd/storage/drivers/utils_cgo.go:292]: I added a separate `ioctl` call using the LOOP_SET_DIRECT_IO operation in the updated commits. The flag is stripped on failure.

**Comment by brauner** [comment]
> So looking into the systemd code linked in the issue, you can find: https://github.com/systemd/systemd/pull/21044/files#diff-e20c8b3ba10a6099dd4cf12f9b95ade0cba34523b5662ac7779b11f92c37da41R482
> 
> Which has the logic to properly handle that flag. Looks like it's slightly more involved than done in this PR as we need to:
> 
> * Make sure that the file backing the loop is open with O_DIRECT as well, if we fail to do that, then we should strip the ioctl flag
> * Do a separate ioctl call to set DIRECT_IO mode on the loop
> * Handle failure of that ioctl as the kernel may not be supported by the kernel
> 
> @brauner does that sound right?

Yes, that does sound correct.

File: lxd/storage/drivers/utils_cgo.go [hunk 4]
```diff
@@ -282,7 +307,6 @@ func PrepareLoopDev(source string, flags int) (*os.File, error) {
 	if loopFd >= 0 {
 		return os.NewFile(uintptr(loopFd), C.GoString((*C.char)(cLoopDev))), nil
 	}
-
```
Comment (tomponline) [inline lxd/storage/drivers/utils_cgo.go:285]: Please can you keep this empty line in. Thanks

File: lxd/storage/drivers/utils_cgo.go [hunk 5]
```diff
@@ -235,6 +258,7 @@ int set_autoclear_loop_device(int fd_loop)
 	return ioctl(fd_loop, LOOP_SET_STATUS64, &lo64);
 }
 
+
```
Comment (stgraber) [inline lxd/storage/drivers/utils_cgo.go:261]: Don't need that line break

File: lxd/storage/drivers/driver_lvm_utils.go (hunk 1 above)
Comment (brauner) [inline lxd/storage/drivers/driver_lvm_utils.go:63]: Where is that argument defined? This should fail to compile given it's the first commit in the series?

Comment (brauner) [reply lxd/storage/drivers/driver_lvm_utils.go:63]: We require that all commits build without an error/warning not just the whole series/branch. To make sure that this works you can do:

```
git rebase origin/master -x "make -j4"
```

Comment (brauner) [reply lxd/storage/drivers/driver_lvm_utils.go:63]: (from your branch that is)

Comment (archevan) [reply lxd/storage/drivers/driver_lvm_utils.go:63]: I have it defined in `utils-cgo.go` on line 291. Should I move it into the driver files?

File: lxd/storage/drivers/utils_cgo.go [hunk 6]
```diff
@@ -348,3 +360,25 @@ func UnsetAutoclearOnLoopDev(loopFd int) error {
 
 	return nil
 }
+
+func SetDirectIOLoopDev(loopFd int) error {
+	ret, err := C.set_direct_io_loop_device(C.int(loopFd))
```
Comment (brauner) [inline lxd/storage/drivers/utils_cgo.go]: This surfaces the errno value, right?
So how about:
```go
if ret < 0 || err != nil {
        if err != nil {
               return err
        }

        return fmt.Errorf....
}

File: lxd/storage/drivers/utils_cgo.go [hunk 7]
```diff
@@ -307,6 +306,7 @@ func PrepareLoopDev(source string, flags int) (*os.File, error) {
 	if loopFd >= 0 {
 		return os.NewFile(uintptr(loopFd), C.GoString((*C.char)(cLoopDev))), nil
 	}
+
```
Comment (brauner) [inline lxd/storage/drivers/utils_cgo.go]: I assume this is a separate commit on purpose requested by another maintainer and is not something that you introduced. If this is just a whitespace fix that already existed make it the first commit of the series. Otherwise don't introduce the whitespace error in the first place. :)

File: lxd/storage/drivers/utils_cgo.go [hunk 8]
```diff
@@ -173,7 +173,7 @@ static int get_unused_loop_dev(char *name_loop)
 		return -1;
 
 	// try to open loop file with O_DIRECT flag, try again without if it fails
-	ret = open(name_loop, O_RDWR | O_CLOEXEC | O_DIRECT)
+	ret = open(name_loop, O_RDWR | O_CLOEXEC | O_DIRECT);
```
Comment (brauner) [inline lxd/storage/drivers/utils_cgo.go:176]: This is something you introduced or that existed before?
If this existed before move this fix to the beginning of the series. If you did introduce it in an earlier commit don't introduce it in the first place. ;)

File: lxd/storage/drivers/utils_cgo.go [hunk 9]
```diff
@@ -360,25 +372,3 @@ func UnsetAutoclearOnLoopDev(loopFd int) error {
...
-func UnsetDirectIOLoopDev(loopFd int) error {
-	ret, err := C.unset_direct_io_loop_device(C.int(loopFd))
-	if ret < 0 {
-		if err != nil {
-			return err
-		}
-		return fmt.Errorf("Failed to unset LO_FLAGS_DIRECT_IO")
-	}
-	return nil
-}
```
Comment (brauner) [inline lxd/storage/drivers/utils_cgo.go]: I'm confused. Why is this deleted and where did it go? If you remove it here why introduce it in the previous commit?

File: lxd/storage/drivers/utils_cgo.go (hunk 6 above)
Comment (archevan) [reply lxd/storage/drivers/utils_cgo.go]: I ended up removing both of the Go functions I added because the `set_direct_io_loop_device` C functions are only ever called from C code.

File: lxd/storage/drivers/utils_cgo.go (hunk 9 above)
Comment (archevan) [reply lxd/storage/drivers/utils_cgo.go]: I initially included these functions just in case I would need to set the `LO_FLAGS_DIRECT_IO` flag from Go but I ended up not needing them after following @stgraber's approach. I'll remove them from the initial commit!

File: lxd/storage/drivers/driver_lvm_utils.go (hunk 1 above)
Comment (archevan) [reply lxd/storage/drivers/driver_lvm_utils.go:63]: I restructured the commits. They should all build without errors now.

File: lxd/storage/drivers/utils_cgo.go [hunk 10]
```diff
@@ -265,6 +265,7 @@ import "C"
 // LoFlagsAutoclear determines whether the loop device will autodestruct on last
 // close.
 const LoFlagsAutoclear int = C.LO_FLAGS_AUTOCLEAR
+const LoFlagsDirectIO int = C.LO_FLAGS_DIRECT_IO
```
Comment (stgraber) [inline lxd/storage/drivers/utils_cgo.go:291]: The commit message for this one is incorrect as it only really adds the const.

Comment (archevan) [reply lxd/storage/drivers/utils_cgo.go:291]: I'll get that changed

Comment (stgraber) [inline lxd/storage/drivers/utils_cgo.go:291]: As this is an exported const, it needs to have a comment similar to the `LoFlagsAutoclear` above it, or it will fail `golint`.

**Comment by stgraber** [comment]
I was about to throw that at Jenkins for testing when I noticed the lack of comment on an exported const, this would make our static analysis immediately fail as it wouldn't be `golint` clean. Once that comment is added, we should be able to run this through CI.

**Comment by stgraber** [comment]
jenkins: test this please
